src/nvim/drawline.c

```diff
@@ -395,6 +398,94 @@ static int get_sign_attrs(buf_T *buf, linenr_T lnum, SignTextAttrs *sattrs, int
   return num_signs;
 }
 
+static void get_statuscol_str(win_T *wp, linenr_T lnum, int row, int startrow, int filler_lines,
+                              int cul_attr, int sign_num_attr, SignTextAttrs *sattrs,
+                              foldinfo_T foldinfo, char_u *extra, statuscol_T *stcp)
+{
+  int sign_idx = 0;
+  bool wrapped = row != startrow + filler_lines;
+  bool use_cul = use_cursor_line_sign(wp, lnum);
+
+  if (!wrapped) {
+    // Set number and fold text and attrs
+    set_vim_var_nr(VV_LNUM, lnum);
+    set_vim_var_nr(VV_RELNUM, labs((long)get_cursor_rel_lnum(wp, lnum)));
+    stcp->num_attr = sign_num_attr ? sign_num_attr
+                     : get_line_number_attr(wp, lnum, row, startrow, filler_lines);
+
+    if (compute_foldcolumn(wp, 0)) {
+      stcp->fold_attr = win_hl_attr(wp, use_cul ? HLF_CLF : HLF_FC);
+      size_t n = fill_foldcolumn((char_u *)stcp->fold_text, wp, foldinfo, lnum);
+      stcp->fold_text[n] = NUL;
+    }
+  }
+  // Set sign text and attrs, empty when wrapped
+  for (; sign_idx < wp->w_scwidth; sign_idx++) {
+    SignTextAttrs *sattr = wrapped ? NULL : sign_get_attr(sign_idx, sattrs, wp->w_scwidth);
+    stcp->sign_text[sign_idx] = sattr && sattr->text ? sattr->text : "  ";
+    stcp->sign_attr[sign_idx] = use_cul && cul_attr ? cul_attr : sattr ? sattr->hl_attr_id : 0;
+  }
+  stcp->sign_text[sign_idx] = NULL;
+  set_vim_var_bool(VV_WRAP, wrapped);
+
+  StlClickRecord *clickrec;
+
+  int width = build_stl_str_hl(wp, stcp->text, MAXPATHL, wp->w_p_stc, "statuscolumn",
+                               OPT_LOCAL, ' ', stcp->width, &stcp->hlrec, &clickrec, stcp);
+
+  // Force a redraw in case of error or when truncated
+  if (*wp->w_p_stc == NUL || (stcp->truncate > 0 && wp->w_nrwidth < 19)) {
+    if (stcp->truncate) {  // Avoid truncating 'statuscolumn'
+      wp->w_nrwidth_width = MIN(19, wp->w_nrwidth_width + stcp->truncate);
+    } else {  // 'statuscolumn' reset due to error
+      wp->w_nrwidth_line_count = 0;
+    }
+    wp->w_redr_statuscol = true;
+    return;
+  }
+  // Reset text pointer and current attr for new line
+  stcp->textp = stcp->text;
+  stcp->cur_attr = stcp->num_attr;
+  stcp->text_len = strlen(stcp->text);
+
+  size_t fill = (size_t)(stcp->width - width);
```

Comment:
The expression contains a subtraction so some form of assert is needed here to avoid underflow. Alternatively use `int` instead of `size_t`.

Reply:
Done (although `stcp->width` is guaranteed to larger than or equal to `width` by `build_stl_str_hl()`